video/out/vo_sixel.c

```diff
@@ -215,42 +202,111 @@ static SIXELSTATUS prepare_dynamic_palette(struct vo *vo)
+    num_cols = (priv->opt_cols > 0)? priv->opt_cols : num_cols;
+
+    // If the pad value is set in between 0 and width/2 - 1, then we
+    // subtract from the detected width. Otherwise, we assume that the width
+    // output must be a integer multiple of num_cols and accordingly set
+    // total_width to be an integer multiple of num_cols. So in case the padding
+    // added by terminal is less than the number of cells in that axis, then rounding
+    // down will take care of correcting the detected width and remove padding.
+    if (priv->opt_width > 0) {
+        // option - set by the user, hard truth
+        total_px_width = priv->opt_width;
+    } else {
+        if (total_px_width <= 0) {
+                // ioctl failed to read terminal width
+                total_px_width = IMG_FALLBACK_DEFAULT_WIDTH;
+        } else {
+            if (priv->opt_pad_x >= 0 && priv->opt_pad_x < total_px_width / 2) {
+                // explicit padding set by the user
+                total_px_width -= (2 * priv->opt_pad_x);
+            } else {
+                // rounded "auto padding"
+                total_px_width = total_px_width / num_cols * num_cols;
+            }
+        }
+    }
 
-    priv->buffer =
-        talloc_array(NULL, uint8_t, depth * priv->width * priv->height);
+    if (priv->opt_height > 0) {
+        total_px_height = priv->opt_height;
+    } else {
+        if (total_px_height <= 0) {
+            total_px_height = IMG_FALLBACK_DEFAULT_HEIGHT;
+        } else {
+            if (priv->opt_pad_y >= 0 && priv->opt_pad_y < total_px_height / 2) {
+                total_px_height -= (2 * priv->opt_pad_y);
+            } else {
+                total_px_height = total_px_height / num_rows * num_rows;
+            }
+        }
+    }
 
-    return 0;
+    // use n-1 rows for height
+    // The last row can't be used for encoding image, because after sixel encode
+    // the terminal moves the cursor to next line below the image, causing the
+    // last line to be empty instead of displaying image data.
+    // TODO: Confirm if the output height must be a multiple of 6, if not, remove
+    // the / 6 * 6 part which is setting the height to be a multiple of 6.
+    vo->dheight = total_px_height * (num_rows - 1) / num_rows / 6 * 6;
+    vo->dwidth  = total_px_width;
+
+    vo_get_src_dst_rects(vo, &priv->src_rect, &priv->dst_rect, &priv->osd);
+
+    // priv->width and priv->height are the width and height of dst_rect
+    // and they are not changed anywhere else outside this function.
+    // It is the sixel image output dimension which is output by libsixel.
+    priv->width  = priv->dst_rect.x1 - priv->dst_rect.x0;
+    priv->height = priv->dst_rect.y1 - priv->dst_rect.y0;
+
+    // top/left values must lie between 1 and num_rows/cols. If it is set, then
+    // the image will be rendered from there and no further centering is done.
+    priv->top  = (priv->opt_top  > 0 && priv->opt_top <= num_rows)?
```

Comment:
I don't think we should check that it's smaller than the number of rows/cols. If the user wants us to go to some valid cell number and it doesn't cause us to divide by 0 or end up in other invalid things, then just do that.

Reply:
Actually, let's check the return value where it's called, and print/log a warning using `MP_WARN` on such case (while still documenting the practical consequences of failure).

Reply:
Okay have removed it.